crates/red_knot_workspace/src/workspace/pyproject/package_name.rs

```diff
@@ -0,0 +1,140 @@
+use serde::{Deserialize, Deserializer, Serialize};
+use std::ops::Deref;
+use thiserror::Error;
+
+/// The normalized name of a package.
+///
+/// Converts the name to lowercase and collapses runs of `-`, `_`, and `.` down to a single `-`.
+/// For example, `---`, `.`, and `__` are all converted to a single `-`.
+///
+/// See: <https://packaging.python.org/en/latest/specifications/name-normalization/>
```

Comment:
Can you not use uv's implementation of this?

Reply:
I used it as a starting point and then changed it.